### PR TITLE
Clean up PupStlCpp11 and Serialize atomic

### PR DESCRIPTION
**Pull request by moxcodes**
## Proposed changes

Removes all of the existing utilities from the file `PupStlCpp11.hpp` that are now in charm 6.10.1, and adds the custom pup for `std::atomic` that charm does not yet have implemented.
Serializing `std::atomic` might seem a bit odd, but we have atomic types in the observers' databox, so in order to have the algorithm function with serializing generic databoxes, it needs to be handled.  Singletons won't migrate during load-balancing, but will need to migrate for checkpointing, so I don't think we can get around it.

### Types of changes:

- [ ] Bugfix
- [x] New feature
- [x] Refactor

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Dependencies
- [ ] #2324 (first two commits here)

### CODE REVIEW

**Comment by moxcodes** [comment]
Squashed in a quick tweak to keep some of the tests that are referenced in snippets from other parts of the code.

**Review by fmahebert — DISMISSED** [review]

File: src/Parallel/PupStlCpp11.hpp [hunk 1]
```diff
@@ -11,4 +11,25 @@
 #include <utility>
 
 namespace PUP {
+
+/// \ingroup ParallelGroup
+/// Serialization of std::atomic for Charm++
+template <typename T>
+inline void pup(PUP::er& p, std::atomic<T>& d) {  // NOLINT
```
Comment (nilsdeppe) [inline src/Parallel/PupStlCpp11.hpp]: This is I think my fault (historically), but we don't need `inline` here and maybe add `noexcept`?

**Review by nilsdeppe — DISMISSED** [review]
Looks good, couple minor, optional thoughts :) Feel free to squash in if you want to make the changes, but don't feel like you have to make them!

**Comment by moxcodes** [comment]
@nilsdeppe Thanks for the review, I've squashed in the suggested changes.
Cheers!

**Review by nilsdeppe — DISMISSED** [review]

**Comment by moxcodes** [comment]
Rebased on develop, no longer dependent.
Cheers!

**Review by nilsdeppe — APPROVED** [review]

**Review by fmahebert — APPROVED** [review]

**Pull request merged**